smart-contracts/contracts/mixins/MixinPurchase.sol

```diff
@@ -106,6 +123,19 @@ contract MixinPurchase is
     {
       onKeyPurchaseHook.onKeyPurchase(msg.sender, _recipient, _referrer, _data, inMemoryKeyPrice, pricePaid);
     }
+
+    // refund gas
+    if (_gasRefundPercentage != 0) {
+      uint toRefund = _gasRefundPercentage * pricePaid / 100;
+      if(tokenAddress != address(0)) {
+        IERC20Upgradeable token = IERC20Upgradeable(tokenAddress);
+        token.transferFrom(address(this), msg.sender, toRefund);
+      } else {
+        (bool success, ) = msg.sender.call{value: toRefund}("");
+        require(success, "Refund failed.");
+      }
+      emit GasRefunded(msg.sender, toRefund, tokenAddress);
```

Comment:
Gas refund as a percentage of the key price means that gas is in ETH while key price will be refunded in any sent token. Does that make sense?

Reply:
Yes and I think that's the only way as we don't know that the lock will have eth to cover the gas... The "submitters" will be able to assess whether it is worth for them to issue the tx based on the conversion rate of a given token.